generate_report.py

```diff
@@ -0,0 +1,44 @@
+import subprocess
+import os
+import json
+from datetime import datetime
+
+# Execute this file to generate a report of deployment latency.  Update max_version to get more recent releases and min_versions to go back further
+max_version = 50
+min_version = 10
+
+if os.path.exists("endpoints.json"):
+    print("Removing endpoints file")
+    os.remove("endpoints.json")
+
+f = open('output.csv', 'w')
+f.write("%s\t%s\t%s\t%s\t%s\t%s\n" % ("version", "merge_date", "last_prod_date", "delta (second)", "delta (hours)", "delta (days)"))
+
+last_prod_date = ""
+for i in range(max_version, min_version , -1):
+    version = "1.%s.0" % i
+    print("Checking version %s" % version)
+    process = subprocess.Popen(["./list.sh", version], stdout=subprocess.PIPE)
```

Comment:
This is fine but am not sure why you don't do it all in python. It would make the way `endpoints.json` is handled a bit cleaner too.

Also if you read the json file in python you don't need min and max as you can sort the semvers and output the latest (or the latest N based on a parameter).

Reply:
Totally.  I just didn't want to re-invent the bits that existed already

Reply:
Here's a short script that reads from the bucket directly and creates a dict of env -> all releases, sorted by date.

```python
import json
from collections import defaultdict
from google.cloud import storage


def main():
    client = storage.Client()
    bucket = client.get_bucket('deployment-tracking')
    blob = bucket.get_blob('endpoints.json')

    data = json.loads(blob.download_as_string())["data"]

    releases = defaultdict(list)
    for release in data:
        releases[release["env"]].append(release)

    for env, env_releases in releases.items():
        env_releases.sort(key=lambda s: s["date"], reverse=True)

    # for each env, print out the two most recent releases
    for env, env_releases in releases.items():
        print(env, env_releases[0], env_releases[1])
    
if __name__ == "__main__":
    # execute only if run as a script
    main()
```

Reply:
This is not yet complete, as we apparently release the same version multiple times (seems like a bug). We probably want the first release of any version.. shouldn't be too hard to filter the lists to do this.

Reply:
This could still be refined but it has the fix to filter by the earliest occurrence of a version, and uses dateutil to print out the time between the last two releases. Using this you can see that it was 14 days and 18 hours between the last two prod releases.
```python
import json
from collections import defaultdict
from google.cloud import storage
import dateutil.parser


def main():
    client = storage.Client()
    bucket = client.get_bucket('deployment-tracking')
    blob = bucket.get_blob('endpoints.json')

    data = json.loads(blob.download_as_string())["data"]

    releases = defaultdict(list)
    for release in data:
        release["date"] = dateutil.parser.parse(release["date"])
        releases[release["env"]].append(release)        

    for env_releases in releases.values():
        env_releases.sort(key=lambda s: s["date"])

    # filter so we have only the earliest release of any version in an env
    for env, env_releases in releases.items():
        seen_versions = set()
        filtered_releases = list()
        for release in env_releases:
            if release["semVer"] not in seen_versions:
                seen_versions.add(release["semVer"])
                filtered_releases.append(release)
        filtered_releases.sort(key=lambda s: s["date"], reverse=True)
        releases[env] = filtered_releases
                
    # for each env, print out the two most recent releases and the time between them
    for env, env_releases in releases.items():
        print(env, env_releases[0], env_releases[1], env_releases[0]["date"] - env_releases[1]["date"])
    
if __name__ == "__main__":
    # execute only if run as a script
    main()
```